src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/OptimizedCompletionList.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Newtonsoft.Json;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion
+{
+    [JsonConverter(typeof(OptimizedCompletionListJsonConverter))]
+    internal class OptimizedCompletionList : CompletionList
```

Comment:
@dibarbet @davidwengier @allisonchou Would Roslyn be interested in doing something like this? I built this with the intention that all "optimization" only exists in this JsonConverter so that the fragility of optimizing completion items doesn't leak elsewhere. Ended up with some pretty significant performance gains:

**Roundtrip op/s improvement:** 1.85x faster
**Serialization op/s improvement:** 2.42x faster
**Deserialization op/s improvement:** 1.59x faster

Reply:
I would be interested in this.  I'm not sure I would want to just straight copy this though, it seems like something that could live in the LSP protocol libraries and could be 'invisible' to the servers.

Reply:
> I would be interested in this. I'm not sure I would want to just straight copy this though, it seems like something that could live in the LSP protocol libraries and could be 'invisible' to the servers.

Totally agree. I tried going that route and was met with push back because of the added complexity. Trying to first see if a few teams adopting this can prove its usefulness to then convince the platform to take it. Also in our case our language server doesn't utilize the VS LSP protocol bits (uses O#'s) so we'd need to have one from the serialization aspect.

Reply:
imho the added complexity is exactly why we'd want this to be written once at the platform layer rather than duplicated across servers.

but we can definitely try it out first by duplicating to see what kind of improvement we get.  If we all get good results perhaps that will be enough evidence.  @allisonchou would you be able to try this out at some point?

Reply:
> imho the added complexity is exactly why we'd want this to be written once at the platform layer rather than duplicated across servers.

100% on the same page lol

Reply:
Yep! I'd be happy to try it out sometime soon.

> Also in our case our language server doesn't utilize the VS LSP protocol bits (uses O#'s) so we'd need to have one from the serialization aspect.

@NTaylorMullen could you possibly expand on this part a little? I think I understood most of the thread up to this point, but is there something that Roslyn will need to do differently because of this if we were to try out something similar to what you've written?

Reply:
> @NTaylorMullen could you possibly expand on this part a little? I think I understood most of the thread up to this point, but is there something that Roslyn will need to do differently because of this if we were to try out something similar to what you've written?

The CompletionList type that our JsonConverter uses is [from O#](https://github.com/OmniSharp/csharp-language-server-protocol/blob/94866ebecb64f4f080c819f1ba4fb734f44fce35/src/Protocol/Features/Document/CompletionFeature.cs#L357). It looks just like VS' CompletionList type but isn't exactly the same. I doubt much would be different in terms of the converter except for some different usings at the top 😄

Reply:
Oh and if you utilize the VS' extension variant of the CompletionList you'll also need to write the logic to serialize those portions as well